File: CPP/Offer/09/QueueUsingStack.cpp
// 
// Created by Insomnia on 18-8-12.
//


#include <iostream>
#include <stack>


using namespace std;
class MyQueue {
public:
    MyQueue() {

    }
    void pushBack(int x) {

        stack<int> tempStack;
        while (!_data.empty()) {
            tempStack.push(_data.top());
            _data.pop();
        }

        tempStack.push(x);
        while (!tempStack.empty()) {
            _data.push(tempStack.top());
            tempStack.pop();
        }
    }

    void pop() {
        _data.pop();
    }
    int front(){
        int x = _data.top();
        return x;
    }
    bool empty() {
        return _data.empty();
    }

private:
    stack<int> _data;
};

int main ()  {

    MyQueue myQueue;
    myQueue.pushBack(1);
    myQueue.pushBack(2);
    myQueue.pushBack(3);
    myQueue.pushBack(4);
    myQueue.pushBack(5);

    myQueue.pop();
    myQueue.pushBack(6);

    while (!myQueue.empty()) {
        cout << myQueue.front() <<  " " ;
        myQueue.pop();
    }
    cout << endl;
    return  0;
}

File: Python3/Offer/09/QueueUsingStack.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Time    : 2018/11/21 4:24 PM
# @Author  : Insomnia
# @Desc    : 用两个栈实现队列
# @File    : QueueUsingStack.py
# @Software: PyCharm


class Solution:
    def __init__(self):
        self.__data= []
        self.__tmp = []


    def push(self, val):
        """
        pycharm 的函数自动注释 是三个引号
        :param val:
        :return:
        """
        if len(self.__data) == 0:
            self.__data.append(val)
        else:
            while self.__data:
                self.__tmp.append(self.__data[-1])
                self.__data.pop()

            self.__data.append(val)

            while self.__tmp:
                self.__data.append(self.__tmp[-1])
                self.__tmp.pop()
        return


    def pop(self):
        if self.__data:
            self.__data.pop()

    def front(self):
        if self.__data:
            return self.__data[-1]
        else:
            return None

if __name__ == '__main__':
    arr = [1, 2, 3]
    sol = Solution()
    for val in arr:
        sol.push(val)
    print(sol.front())
    sol.pop()
    print(sol.front())
    sol.pop()
    print(sol.front())
    sol.pop()
    print(sol.front())
    sol.pop()
    print(sol.front())


